### PR TITLE
Iox #337 create periodic task

### LINKED ISSUE #337: Deprecate posix timer
## Brief feature description

Deprecate posix timer from iceoryx code base.

## Detailed information

Posix timer has several issues (concurrency, truncation) and is pretty heavy weight and a bit slow. As iceorxy needs to be very responsive the posix timer should be removed and replaced with a lightweight stopwatch.

This will also close #190 as it'll be obsolete.

Tasks
- [ ] create `DeadlineTimer` or similar and replace this part of the `PosixTimer` where applicable
- [ ] create a small wrapper around a std::thread which can be used for the keep alive message
    - replace the keep alive message from a `std::chrono` timestamp with a `units::Duration` from a `timespec` initialized with `clock_gettime` with `CLOCK_REALTIME`

### PR DESCRIPTION
**Pull request by elBoberido**
## Pre-Review Checklist for the PR Author

1. [x] Branch follows the naming format (`iox-#123-this-is-a-branch`)
1. [x] Commits messages are according to this [guideline][commit-guidelines]
    - [x] Commit messages have the issue ID (`iox-#123 commit text`)
    - [x] Commit messages are signed (`git commit -s`)
    - [x] Commit author matches [Eclipse Contributor Agreement][eca] (and ECA is signed)
1. [x] Update the PR title
   - Follow the same conventions as for commit messages
   - Link to the relevant issue
1. [x] Relevant issues are linked
1. [x] Add sensible notes for the reviewer
1. [x] All checks have passed
1. [x] Assign PR to reviewer

[commit-guidelines]: https://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html
[eca]: http://www.eclipse.org/legal/ECA.php

## Notes for Reviewer
This PR creates a `PeriodicTask` class which can be used to, well, execute periodic tasks. It also replaces the current posix::Timer to send the keep alive message. Once the periodic::Timer is refactored, it could be used as base for PeridodicTask

## Checklist for the PR Reviewer

- [ ] Commits are properly organized and messages are according to the guideline
- [ ] Code according to our coding style and naming conventions
- [ ] Unit tests have been written for new behavior
- [ ] Public API changes are documented via doxygen
- [ ] Copyright owner are updated in the changed files
- [ ] PR title describes the changes

## Post-review Checklist for the PR Author

1. [ ] All open points are addressed and tracked via issues

## Post-review Checklist for the Eclipse Committer

1. [ ] All checkboxes in the PR checklist are checked or crossed out
1. [ ] Merge

## References

- Partly closes #337

### CODE REVIEW

File: iceoryx_utils/include/iceoryx_utils/internal/concurrent/periodic_task.hpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+namespace concurrent
+{
+/// @brief This class periodically executes a callable specified by the template parameter.
+///        This can be a struct with a `operator()()` overload, a `cxx::function_ref<void()>` or `std::fuction<void()>`
+/// @tparam T is a callable type without function parameters
+template <typename T>
+class PeriodicTask
+{
+  public:
+    /// @brief Creates a periodic task by spawning a thread. The specified callable is executed immediatelly on creatian
```
Comment (budrus) [inline iceoryx_utils/include/iceoryx_utils/internal/concurrent/periodic_task.hpp:36]: typo creation

File: iceoryx_utils/test/moduletests/test_concurrent_periodic_task.cpp [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+TEST_F(PeriodicTask_test, PeriodicTaskWithObjectWithDefaultConstructor)
+{
+    {
+        using namespace iox::units::duration_literals;
+        concurrent::PeriodicTask<PeriodicTaskTestType> sut("Test", 10_ms);
+
+        std::this_thread::sleep_for(std::chrono::milliseconds(100));
+    }
+
+    EXPECT_THAT(PeriodicTaskTestType::callCounter, AllOf(Gt(5U), Lt(15U)));
```
Comment (budrus) [inline iceoryx_utils/test/moduletests/test_concurrent_periodic_task.cpp:107]: I see the next timing tests coming....

**Review by budrus — APPROVED** [review]

File: iceoryx_utils/test/moduletests/test_concurrent_periodic_task.cpp [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        ++callCounter;
+    }
+
+    static uint64_t callCounter;
+};
+
+uint64_t PeriodicTaskTestType::callCounter{0};
+
+
+using CallableTypes =
```
[the remark this answers is not in the document]
Comment (elBoberido) [reply iceoryx_utils/test/moduletests/test_concurrent_periodic_task.cpp]: nope

File: iceoryx_utils/test/moduletests/test_concurrent_periodic_task.cpp (hunk 2 above)
Comment (elBoberido) [reply iceoryx_utils/test/moduletests/test_concurrent_periodic_task.cpp:153]: hahaha, you're right. The `PeriodicTask` uses a `Semaphore` with `timedWait`. And the `Semaphore` timing tests are disabled on mac. So probably I need to make this a timing test

File: iceoryx_utils/test/moduletests/test_concurrent_periodic_task.cpp [hunk 4]
```diff
@@ -0,0 +1,218 @@
...
+
+    static void increment()
+    {
+        ++callCounter;
+    }
+
+    static uint64_t callCounter;
+};
+
+uint64_t PeriodicTaskTestType::callCounter{0};
```
Comment (dkroenke) [inline iceoryx_utils/test/moduletests/test_concurrent_periodic_task.cpp:55]: :eyes:

File: iceoryx_utils/platform/mac/source/semaphore.cpp [hunk 5]
```diff
@@ -145,6 +145,11 @@ int iox_sem_timedwait(iox_sem_t* sem, const struct timespec* abs_timeout)
     if (sem->m_hasPosixHandle)
     {
         int tryWaitCall = sem_trywait(sem->m_handle.posix);
+        constexpr int ETIMEDOUT_PLUS_256 = ETIMEDOUT + 256;
+        if (errno == ETIMEDOUT_PLUS_256)
+        {
+            errno &= 0xFF;
+        }
```
Comment (elBoberido) [inline iceoryx_utils/platform/mac/source/semaphore.cpp:152]: @elfenpiff do you know why this works

**Review by dkroenke — APPROVED** [review]

**Review by dkroenke — DISMISSED** [review]

**Review by mossmaurice — APPROVED** [review]

**Review by dkroenke — APPROVED** [review]

**Review by mossmaurice — APPROVED** [review]

File: iceoryx_utils/test/moduletests/test_concurrent_periodic_task.cpp (hunk 4 above)
Comment (elBoberido) [reply iceoryx_utils/test/moduletests/test_concurrent_periodic_task.cpp:66]: whoopsi

**Pull request merged**